src/codegen/ast_interpreter.cpp

```diff
@@ -429,199 +527,266 @@ void ASTInterpreter::doStore(AST_expr* node, Value value) {
-                    continue;
+    if (jit) {
+        if (backedge)
+            jit->emitOSRPoint(node);
+        jit->emitJump(node->target);
+        finishJITing(node->target);
+    }
 
-                found_entry = p.first;
-            }
+    if (backedge)
+        ++edgecount;
 
-            std::map<InternedString, Box*> sorted_symbol_table;
+    if (ENABLE_BASELINEJIT && backedge && edgecount == OSR_THRESHOLD_INTERPRETER && !jit && !node->target->code)
+        startJITing(node->target);
 
-            for (auto& name : phis->definedness.getDefinedNamesAtEnd(current_block)) {
-                auto it = sym_table.find(name);
-                if (!liveness->isLiveAtEnd(name, current_block))
-                    continue;
+    if (backedge && edgecount == OSR_THRESHOLD_BASELINE) {
+        Box* rtn = doOSR(node);
+        if (rtn)
+            return Value(rtn, NULL);
+    }
 
-                if (phis->isPotentiallyUndefinedAfter(name, current_block)) {
-                    bool is_defined = it != sym_table.end();
-                    // TODO only mangle once
-                    sorted_symbol_table[getIsDefinedName(name, source_info->getInternedStrings())] = (Box*)is_defined;
-                    if (is_defined)
-                        assert(sym_table.getMapped(it->second) != NULL);
-                    sorted_symbol_table[name] = is_defined ? sym_table.getMapped(it->second) : NULL;
-                } else {
-                    ASSERT(it != sym_table.end(), "%s", name.c_str());
-                    sorted_symbol_table[it->first] = sym_table.getMapped(it->second);
-                }
-            }
+    next_block = node->target;
+    return Value();
+}
 
-            // Manually free these here, since we might not return from this scope for a long time.
-            liveness.reset(nullptr);
-            phis.reset(nullptr);
-
-            // LLVM has a limit on the number of operands a machine instruction can have (~255),
-            // in order to not hit the limit with the patchpoints cancel OSR when we have a high number of symbols.
-            if (sorted_symbol_table.size() > 225) {
-                static StatCounter times_osr_cancel("num_osr_cancel_too_many_syms");
-                times_osr_cancel.log();
-                next_block = node->target;
-                return Value();
-            }
+Box* ASTInterpreter::doOSR(AST_Jump* node) {
+    bool can_osr = ENABLE_OSR && !FORCE_INTERPRETER && source_info->scoping->areGlobalsFromModule();
+    if (!can_osr)
+        return NULL;
+
+    static StatCounter ast_osrs("num_ast_osrs");
+    ast_osrs.log();
 
-            if (generator)
-                sorted_symbol_table[source_info->getInternedStrings().get(PASSED_GENERATOR_NAME)] = generator;
+    // TODO: we will immediately want the liveness info again in the jit, we should pass
+    // it through.
+    std::unique_ptr<LivenessAnalysis> liveness = computeLivenessInfo(source_info->cfg);
```

Comment:
Could this be getLiveness()?  I think the call to computeLivenessInfo() in irgen.cpp could also be changed to access the cached version.